### DISCUSSION TITLE
uri path variable question

### DISCUSSION THREAD

**Discussion by esundberg** [fastapi/fastapi#7584] [Questions] [answered]
### First check

* [ x ] I used the GitHub search to find a similar issue and didn't find it.
* [ x ] I searched the FastAPI documentation, with the integrated search.
* [ x ] I already searched in Google "How to X in FastAPI" and didn't find any information.

### Description

 /cars and /cars/{uuid} use the same decorator/function

### Additional context

Looking to to see if there is a way to have /cars and /cars/{uuid} use the same decorator/function and not create duplicate code? Is there a way to register the URI with an optionial {uuid} in the path? I can't find anything in the docs on this, everything that I saw was about URI key/value pairs after the ? in the URI

Current Way - That works
```python
class Cars_Get(BaseModel):
    uuid: UUID = Field(None, title="UUID")
    message: str

@app.get("/cars", response_model=List[Cars_Get])'
async def missedcall():
	return [{"uuid": 'd50eae76-8e0e-11ea-bc55-0242ac130003', "message": "all cars"}]

@app.get("/cars/{uuid}", response_model=Cars_Get)
async def missedcall(uuid: UUID = None):
	    return {"uuid": uuid,"message": "Single car" }
```

Tried the following but get
```
INFO:     *.*.*.74:43020 - "GET /cars HTTP/1.1" 404 Not Found
```

```python
@app.get("/cars/{uuid}", response_model=Cars_Get)
async def missedcall(uuid: UUID = None):
    if not uuid:
	    return [{"uuid": 'd50eae76-8e0e-11ea-bc55-0242ac130003', "message": "all cars"}]
    else:
	    return {"uuid": uuid,"message": "Single car" }

```

Tried this way but I get a 307 Temporary Redirect, but it works. Not sure if it's really valid

```
INFO:     73.22.78.74:42765 - "GET /cars/ HTTP/1.1" 307 Temporary Redirect
INFO:     73.22.78.74:42765 - "GET /cars HTTP/1.1" 200 OK

INFO:     73.22.78.74:42964 - "GET /cars/d50eae76-8e0e-11ea-bc55-0242ac130003 HTTP/1.1" 200 OK
```

```python
class Cars_Get(BaseModel):
    uuid: UUID = Field(None, title="UUID")
    message: str

@app.get("/cars", response_model=List[Cars_Get])
@app.get("/cars/{uuid}", response_model=Cars_Get)
async def missedcall(uuid: UUID = None):
    if not uuid:
	    return [{"uuid": 'd50eae76-8e0e-11ea-bc55-0242ac130003', "message": "all cars"}]
    else:
	    return {"uuid": uuid,"message": "Single car" }
```

**Comment by retnikt** [comment]
The way you tried is perfectly valid. To avoid getting the redirect replace `/cars` with `/cars/`.

PS: put ``` before and after your code to keep it formatted properly. E.g.:

    This is normal text
    
    ```
    this.is_some(code)
    ```
    
    This is normal text again
-----
This is normal text

```
this.is_some(code)
```

This is normal text again

**Comment by esundberg** [comment]
Thanks for your help

**Comment by tiangolo** [comment]
Thanks for the help here @retnikt ! :clap: :bow: 

Thanks for reporting back and closing the issue @esundberg  :+1:

**Comment by YuriiMotov** [comment, marked as answer]
Although the code example with two path decorators works (you can add third to avoid redirects), it's not good since the produced openapi schema is not as you probably expect (your `uuid` Path parameter turns into Query parameter for `/cars` path).

I would just create several separate paths and implement common logic as a function to avoid code duplication:

```python
from typing import Optional
from uuid import UUID

from fastapi import FastAPI

app = FastAPI()


def get_car(uuid: Optional[UUID] = None):
    if not uuid:
        return [{"uuid": "d50eae76-8e0e-11ea-bc55-0242ac130003", "message": "all cars"}]
    return {"uuid": uuid, "message": "Single car"}


@app.get("/cars")
async def get_cars():
    return get_car()


@app.get("/cars/{uuid}")
async def get_car_by_uuid(uuid: UUID):
    return get_car(uuid)

```
